### PR TITLE
[FLINK-8686] [sql-client] Improve basic embedded SQL client

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This PR covers the following minor improvements:

- Configure JVM heap size
- "The input is invalid please check it again." => add allowed range
- Add more logging instead of swallowing exceptions
- List properties in error message about missing TS factory sorted by name
- Add switch to show full stacktraces of exceptions (solved through logging)
- Minor improvements

## Brief change log

Various changes.

## Verifying this change

Manually verified and existing tests adapted.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

**Comment by twalthr** [comment]
@fhueske @walterddr @suez1224 I added the possibility to make the JVM heap size configurable. After a short offline discussion with @StephanEwen, I'm not sure if we should add such configuration to the flink-conf.yaml or in a separate configuration file. I don't know which configuration we will need in the future outside of the SQL Client environment files. I don't know if it is worth to have a separate file just for JVM options. Or we hard code it into the sql-client.sh script that can be modified by a variable. What do you think?

**Comment by walterddr** [comment]
Hi @twalthr, having JVM heap size configurable is definitely a great benefit. Just to clarify, this is only changing the Client JVM heap size, correct? I am assuming this is mainly for manipulating large size data when retrieving results from the query. 

I think adding it to `flink-conf.yaml` should be fine. There are already specific configurations commented out for components like Kerberos, ZK, etc. which makes adding a specific conf to sql-client less disturbing. We can always break it out to another file later when extra configuration files are needed for sql-client in the future. 

What do you guys think?

**Comment by suez1224** [comment]
Hi @twalthr ,

IMO, it's better to seperate the configuration of SQL client from the configuration of the Flink job. From the user's perspective, many of them probably do not need to know or use SQL client, putting a SQL client section in flink-conf.yaml will just confuse the users because they might think that's part of their flink job configuration. 

I suggest, as you already mentioned, we can either 1) add some variable in the sql-client.sh to allow the advanced users to modify the JVM ARGS within the script or 2) add a sql-client-conf.yaml to allow advanced user to edit it.

Assuming that it is uncommon to modify the JVM options of the sql client, I think both options are fine, but I slightly prefer option 2 since sql-client.sh is not supposed to modified by the users.

File: flink-libraries/flink-sql-client/bin/sql-client.sh [hunk 1]
```diff
@@ -44,20 +44,32 @@ bin=`dirname "$target"`
...
 
-export FLINK_ROOT_DIR
-export FLINK_CONF_DIR
+if [[ ! ${FLINK_SCC_HEAP} =~ ${IS_NUMBER} ]] || [[ "${FLINK_SCC_HEAP}" -lt "0" ]]; then
+    echo "[ERROR] Configured SQL Client CLI JVM heap size is not a number. Please set '${KEY_SCC_MEM_SIZE}' in ${FLINK_CONF_FILE}."
+    exit 1
+fi
+
+if [ "${FLINK_SCC_HEAP}" -gt "0" ]; then
+    export JVM_ARGS="$JVM_ARGS -Xms"$FLINK_SCC_HEAP"m -Xmx"$FLINK_SCC_HEAP"m"
```
Comment (suez1224) [inline flink-libraries/flink-sql-client/bin/sql-client.sh]: Do we need to inherit $JVM_ARGS here? The JVM_ARGS from config.sh is used for the JVMs of JobManager and TaskManagers as suggested in the comment. The sql client might inherit some unwanted JVM options.

File: flink-dist/src/main/flink-bin/bin/config.sh [hunk 2]
```diff
@@ -95,6 +95,8 @@ DEFAULT_ENV_LOG_MAX=5                               # Maximum number of old log
 DEFAULT_ENV_JAVA_OPTS=""                            # Optional JVM args
 DEFAULT_ENV_JAVA_OPTS_JM=""                         # Optional JVM args (JobManager)
 DEFAULT_ENV_JAVA_OPTS_TM=""                         # Optional JVM args (TaskManager)
+DEFAULT_ENV_JAVA_OPTS_SCC=""                        # Optional JVM args (SQL Client CLI)
+DEFAULT_ENV_JAVA_OPTS_SCG=""                        # Optional JVM args (SQL Client Gateway)
```
Comment (fhueske) [inline flink-dist/src/main/flink-bin/bin/config.sh]: The Gateway parameter is not used yet. I think we can added when we need it

File: flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java [hunk 3]
```diff
@@ -91,6 +92,22 @@ public Deployment getDeployment() {
 		return deployment;
 	}
 
+	public String explain() {
+		final StringBuilder sb = new StringBuilder();
+		sb.append("===================== Tables =====================\n");
+		tables.forEach((name, table) -> {
+			sb.append("- name: ").append(name).append("\n");
+			final DescriptorProperties props = new DescriptorProperties(true);
+			table.addProperties(props);
```
Comment (fhueske) [inline flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java]: Should we make a copy of `table` before modifying it?

File: flink-dist/src/main/resources/flink-conf.yaml [hunk 4]
```diff
@@ -229,17 +229,26 @@ web.port: 8081
...
+# historyserver.archive.fs.refresh-interval: 10000
+
+#==============================================================================
+# SQL Client
+#==============================================================================
+
+# The SQL Client CLI can be started via bin/sql-client.sh embedded
+
+# The heap size for the SQL Client CLI JVM
+# sqlclient.cli.heap.mb: 1024
```
Comment (fhueske) [inline flink-dist/src/main/resources/flink-conf.yaml]: Regarding the question whether to add this property of the `flink-conf.yaml` or a CLI client specific file, are there other (existing) properties in the `flink-conf.yaml` file that are used for the CLI client, besides the connection info to the cluster?

Comment (twalthr) [reply flink-dist/src/main/resources/flink-conf.yaml]: If the user wants to specify global JVM options for all Flink scripts, this also happens in the `flink-conf.yaml`.

File: flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java [hunk 5]
```diff
@@ -91,6 +92,22 @@ public Deployment getDeployment() {
 		return deployment;
 	}
 
+	public String explain() {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java]: I will just use the `toString` method instead.

File: flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java (hunk 3 above)
Comment (twalthr) [reply flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/config/Environment.java]: The properties are added to `props` not `table`. Maybe the name is a bit misleading.

**Comment by twalthr** [comment]
Thank you for the feedback. I will merge the first 5 commits and will open a new PR for an additional configuration file.

File: flink-dist/src/main/resources/flink-conf.yaml (hunk 4 above)
Comment (fhueske) [reply flink-dist/src/main/resources/flink-conf.yaml]: OK. But there's another discussion thread whether it is useful to share properties across JM/TM JVMs and client processes.

**Pull request closed without merging**